### ISSUE TITLE
evaluation.metrics.mean_iou() computes recall (sensitivity) instead of IoU (Jaccard)

### ISSUE THREAD

**Issue by FlorinAndrei** [open-mmlab/mmsegmentation#2655]
Thanks for your error report and we appreciate it a lot.

**Checklist**

1. I have searched related issues but cannot get the expected help.
2. The bug has not been fixed in the latest version.

**Describe the bug**
For a project, I've trained a SegFormer model to segment medical images. I've tried to maximize IoU, as the main performance indicator. I've used the Huggingface Evaluate library for it.

Turns out, Evaluate has a bug where it calculates recall (sensitivity) instead of IoU (Jaccard).

https://github.com/huggingface/evaluate/issues/421

They told me they use the same logic that mmsegmentation uses. I've tested it, and sure enough, mmsegmentation has the exact same problem. What it calls "mean_iou" is actually mean recall (mean sensitivity). The bug is in the `intersect_and_union()` function.

https://github.com/open-mmlab/mmsegmentation/blob/70477d21adfb4caaf2e35c8e6d88b9ee740486d6/mmseg/core/evaluation/metrics.py#L26

The `ignore_index` variable wrongfully creates a mask based on pixel values in the labels (`gt_seg_maps`). The mask is then applied to both prediction and label. Portions of the prediction and the label are then ignored from the evaluation, based on the pixel value. This is incorrect.

To correctly calculate IoU, the entirety of the prediction and the label frames must be investigated to calculate intersection and union. Your code fails to calculate the union correctly, by ignoring chunks of the frames.

If you want to exclude a certain index value from the mean IoU, simply skip that pixel value in the calculations, and skip it in the formula for the mean.

Here's a code example I put together. It is incomplete, the comments are wrong, and I've only tested it with two classes: background, plus one object class. But it calculates IoU correctly. It also calculates recall correctly - and it turns out it's equal to what you claim to be "IoU".

https://gist.github.com/FlorinAndrei/da9ab770b16bfc671075d04a030f548b

**Reproduction**

1. What command or script did you run?

```python
import numpy as np
import evaluate
from matplotlib import pyplot as plt

pred = np.zeros((200, 200), dtype=np.int32)
mask = np.zeros((200, 200), dtype=np.int32)

pred[100:, ...] = 1
mask[..., 100:] = 1

f, axs = plt.subplots(1, 2)
axs[0].imshow(pred)
axs[1].imshow(mask)
```

![output](https://user-images.githubusercontent.com/901867/221646936-99f8964d-e027-4ecd-8fd3-7eaa30eb5b20.png)

```python
from mmseg.core.evaluation.metrics import mean_iou as mmseg_miou

mmseg_miou(
    results=[pred],
    gt_seg_maps=[mask],
    num_classes=2,
    ignore_index=0,
    reduce_zero_label=False,
)
```

According to the definition of the IoU / Jaccard index, IoU for the above case for the yellow pixel class should be 0.333. This is because:

- the intersection is the bottom-right corner, 100x100 pixels
- the union is three corners, 100x100 each

Your code ignores the left half of both frames, due to the way `ignore_index` is used. This leads to an IoU value of 0.5, which is clearly wrong. In my example, 0.5 is the value of recall (sensitivity), which is what your code calculates.

IoU = TP / (TP + FP + FN)

recall = TP / (TP + FN)

https://en.wikipedia.org/wiki/Jaccard_index

![Intersection_over_Union_-_visual_equation](https://user-images.githubusercontent.com/901867/221747689-1c6f57eb-fd06-4625-87c0-fa77c3650706.png)

Because recall is typically larger than IoU, your code artificially inflates the IoU value it reports. This is what led me to pursue this investigation - I was getting IoU values from an image segmentation model which were too good to be true.

**Environment**

```
sys.platform: linux
Python: 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0]
CUDA available: True
GPU 0: NVIDIA GeForce RTX 3090
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 11.7, V11.7.99
GCC: x86_64-linux-gnu-gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
PyTorch: 1.13.0+cu117
PyTorch compiling details: PyTorch built with:
  - GCC 9.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.6.0 (Git Hash 52b5f107dd9cf10910aaa19cb47f3abf9b349815)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.7
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86
  - CuDNN 8.5
  - Magma 2.6.1
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.7, CUDNN_VERSION=8.5.0, CXX_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/c++, CXX_FLAGS= -fabi-version=11 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wunused-local-typedefs -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.13.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF,

TorchVision: 0.14.0+cu117
OpenCV: 4.6.0
MMCV: 1.7.1
MMCV Compiler: GCC 9.3
MMCV CUDA Compiler: 11.7
MMSegmentation: 0.30.0+70477d2
```

**Comment by MeowZheng** [comment]
Thank you very much for pointing this out, and sorry for the late reply.

However, I want to explain why we do these when calculated mean iou
>The ignore_index variable wrongfully creates a mask based on pixel values in the labels (gt_seg_maps). The mask is then 
> applied to both prediction and label. 

When model training, the pixels with `ignored_index` are also ignored in the loss function. 
https://github.com/open-mmlab/mmsegmentation/blob/c26915450564e9863980f40c2105c1b2a77a5f07/mmseg/models/losses/cross_entropy_loss.py#L260
In other words, we don't impose any constraints on the output of this part of the pixels, i.e. we don't care about this part of the pixels, and they don't join in model training.

Therefore, back to the example  
> ![output](https://user-images.githubusercontent.com/901867/221646936-99f8964d-e027-4ecd-8fd3-7eaa30eb5b20.png)

the left half pixels are never joined in model training, so I think it should also be ignored when evaluating, and this left-bottom corner is not FP.

The above are our thoughts, if there is anything wrong, just point it out, and we will appreciate you so much.

I saw you are working on a project, would you like to share the setting of your project, like what's the number of classes?
> For a project, I've trained a SegFormer model to segment medical images.

I guess you might set `reduce_zero_label=True`, and turn the 2 classes of datasets became 1 classes, as the pixel with label  indel 0 didn't join the training

**Comment by DavidFM43** [comment]
Hi, I have a question regarding the example. Shouldn't the mean IoU be 0.5 since we are excluding index 0? It seems that the current calculation doesn't reflect this:
```
import torch
from mmseg.evaluation.metrics.iou_metric import IoUMetric


gt_sem_seg_data = torch.zeros((200, 200), dtype=torch.int64)
pred_sem_seg_data = torch.zeros((200, 200), dtype=torch.int64)
gt_sem_seg_data[..., 100:] = 1
pred_sem_seg_data[100:, ...] = 1


data_sample = {"gt_sem_seg": {"data": gt_sem_seg_data}, "pred_sem_seg": {"data": pred_sem_seg_data}}
data_samples = [data_sample]

iou = IoUMetric(ignore_index=0)
iou.dataset_meta = dict(classes=["black", "white"])
iou.process([0] * len(data_samples), data_samples)
iou.evaluate(1)

## Output
# | Class | IoU  | Acc  |
# +-------+------+------+
# | black | 0.0  | nan  |
# | white | 50.0 | 50.0 |
# +-------+------+------+
# {'aAcc': 50.0, 'mIoU': 25.0, 'mAcc': 50.0}
```
Looking into the code, I found this happens because the mean doesn't actually ignore the index but only ignore Nan's:

https://github.com/open-mmlab/mmsegmentation/blob/00790766aff22bd6470dbbd9e89ea40685008395/mmseg/evaluation/metrics/iou_metric.py#L136

And a Nan only appears(in the position of the ignored index) if the prediction doesn't contain any pixel labeled with the ignored index. Is there a reason for this? I guess the prediction shouldn't contain a pixel labeled with the ignored index in the first place, but is this a reasonable assumption in general?

Thanks.
ping @MeowZheng
